### PR TITLE
fix: Button disabled style with link or href case

### LINKED ISSUE #42997: button type = 'link' 并且禁用的情况下 样式问题
### Reproduction link

https://codesandbox.io/s/antd-reproduction-template-forked-pcwx68?file=/index.js

### Steps to reproduce

link 后面加个 href

### What is expected?

正常的link禁用样式

### What is actually happening?

正常button禁用的样式

| Environment | Info |
| --- | --- |
| antd | 5.6.1 |
| React | 18.0.0 |
| System | mac |
| Browser | 113.0.5672.126 |

### PR DESCRIPTION
**Pull request by BoyYangzai**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link
- fix https://github.com/ant-design/ant-design/issues/42997

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |     fix: Button disabled style with link or href case      |
| 🇨🇳 Chinese |     修复 Button 在有 link or href 属性时禁用样式错误      |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at edd3101</samp>

This pull request improves the button and cascader components by refactoring some code, fixing some bugs, adding some features, and updating some documentation and dependencies. It mainly affects the files `components/button/button.tsx`, `components/cascader/demo/multiple.tsx`, `components/button/style/index.ts`, `components/cascader/__tests__/index.test.tsx`, `components/cascader/demo/multiple.md`, and `package.json`.

### <samp>🤖 Generated by Copilot at edd3101</samp>

*  Import and use `IconWrapper` component to wrap the icon prop of the `button` component and apply some styles and logic to it ([link](https://github.com/ant-design/ant-design/pull/43021/files?diff=unified&w=0#diff-463cd38ab465e1f3b90463a403d8e8399f926361436eaa9adb8a76a8bf9d4140R21), [link](https://github.com/ant-design/ant-design/pull/43021/files?diff=unified&w=0#diff-463cd38ab465e1f3b90463a403d8e8399f926361436eaa9adb8a76a8bf9d4140R220-R221))
* Remove duplicate import of `IconWrapper` component from `button.tsx` ([link](https://github.com/ant-design/ant-design/pull/43021/files?diff=unified&w=0#diff-463cd38ab465e1f3b90463a403d8e8399f926361436eaa9adb8a76a8bf9d4140L26))
* Refactor logic for applying CSS classes to the `button` component based on its props using `isDisabledClass` and `isTypeClass` variables ([link](https://github.com/ant-design/ant-design/pull/43021/files?diff=unified&w=0#diff-463cd38ab465e1f3b90463a403d8e8399f926361436eaa9adb8a76a8bf9d4140R220-R221), [link](https://github.com/ant-design/ant-design/pull/43021/files?diff=unified&w=0#diff-463cd38ab465e1f3b90463a403d8e8399f926361436eaa9adb8a76a8bf9d4140L226-R228), [link](https://github.com/ant-design/ant-design/pull/43021/files?diff=unified&w=0#diff-463cd38ab465e1f3b90463a403d8e8399f926361436eaa9adb8a76a8bf9d4140L235-R237))
* Add `mergedDisabled` prop to the native button element rendered by the `button` component to ensure it is disabled when the component is disabled ([link](https://github.com/ant-design/ant-design/pull/43021/files?diff=unified&w=0#diff-463cd38ab465e1f3b90463a403d8e8399f926361436eaa9adb8a76a8bf9d4140R259))
* Add new style rule to the `genPureDisabledButtonStyle` function to set the hover color of the button text to the disabled text color when the button is disabled and has a ghost or dashed type ([link](https://github.com/ant-design/ant-design/pull/43021/files?diff=unified&w=0#diff-e8f947f22fc66308c74c500faf1cfeab5e8eed33d2ab4a27718770da99224693R194-R196))
* Add new feature to the `cascader` component to allow disabling the checkbox of some options using the `disableCheckbox` prop ([link](https://github.com/ant-design/ant-design/pull/43021/files?diff=unified&w=0#diff-5512a04d580043e1237e0df7b627d4743e2e78d2fe596301b981799769338fadR684-R717), [link](https://github.com/ant-design/ant-design/pull/43021/files?diff=unified&w=0#diff-f678e6b02bb8dc90b00293dfd22c410bd961f705317b3bdbef86257c4adc6c89L3-R7), [link](https://github.com/ant-design/ant-design/pull/43021/files?diff=unified&w=0#diff-87981d0bd5784f0e51afb68e555122314f3c1f23d94a9269a1adc39c20f568b2R8), [link](https://github.com/ant-design/ant-design/pull/43021/files?diff=unified&w=0#diff-87981d0bd5784f0e51afb68e555122314f3c1f23d94a9269a1adc39c20f568b2R30))
* Update the version of the `rc-cascader` dependency in the `package.json` file to use the latest features and fixes of the low-level component ([link](https://github.com/ant-design/ant-design/pull/43021/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L123-R123))
* Reorder the imports of the `multiple.tsx` demo file for the `cascader` component to move the React import to the top ([link](https://github.com/ant-design/ant-design/pull/43021/files?diff=unified&w=0#diff-87981d0bd5784f0e51afb68e555122314f3c1f23d94a9269a1adc39c20f568b2L1-R2))

### CODE REVIEW

**Comment by BoyYangzai** [comment]
这个 pr 稍等一下，晚上补充下 test case

File: components/button/button.tsx [hunk 1]
```diff
@@ -262,6 +263,7 @@ const InternalButton: React.ForwardRefRenderFunction<
   if (linkButtonRestProps.href !== undefined) {
     return wrapSSR(
       <a
+        disabled={mergedDisabled}
```
[the remark this answers is not in the document]
Comment (MadCcc) [reply components/button/button.tsx:266]: 当时加 className 应该就是兼容这个情况的

[the remark this answers is not in the document]
Comment (BoyYangzai) [reply components/button/button.tsx:266]: 是的 没有的 加的自定义属性 用来匹配 [disabled] 选择器的样式

[the remark this answers is not in the document]
Comment (BoyYangzai) [reply components/button/button.tsx:266]: 目前 disabled 这个的样式在正常状态都都是来自 [disabled] 这个选择器

File: components/button/button.tsx [hunk 2]
```diff
@@ -218,7 +218,8 @@ const InternalButton: React.ForwardRefRenderFunction<
 
   const linkButtonRestProps = omit(rest as ButtonProps & { navigate: any }, ['navigate']);
 
-  const hrefAndDisabled = linkButtonRestProps.href !== undefined && mergedDisabled;
+  const hrefAndDisabled = linkButtonRestProps.href && mergedDisabled;
```
Comment (Wxh16144) [inline components/button/button.tsx:221]: 后面逻辑都调整了,  命名还叫 `hrefAndDisabled` 是不是会有阅读歧义

Comment (BoyYangzai) [reply components/button/button.tsx:221]: fine

**Comment by afc163** [comment]
混合了无关的 commits

**Comment by Wxh16144** [comment]
> 混合了无关的 commits

这个应该是合并造成的吧🤔, 最终 file change 是对的. 不过在 Squash Merge 的时候可以考虑删除掉

**Comment by afc163** [comment]
这个分支大概不是从 master 中切出来的。

**Review by afc163 — CHANGES_REQUESTED** [review]
混合了无关的 commits，请清理

**Pull request closed without merging**